backup: Export tenant backup schedule metric

edwardguo-crl opened the pull request:
Export Cloud tenant backup schedule completion metric with tenant id tag.
    
Issue: #141167

Epic: None
   
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/141156)
<!-- Reviewable:end -->

edwardguo-crl commented:
Quick question: Is there a way I can test the exported metric before running it in staging? @msbutler

kev-cao commented on pkg/backup/schedule_exec.go:
```diff
@@ -578,6 +584,12 @@ func init() {
 					}),
+					RpoTenantMetric: NewExportedGaugeVec(metric.Metadata{
+						Name:        "schedules.BACKUP.tenant-last-completed-time",
+						Help:        "The unix timestamp of the most recently completed backup by a tenant schedule specified as maintaining this metric",
```
This isn't a "tenant schedule" but rather a schedule for backing up a tenant. So the schedule still exists on the host cluster, but backs up the tenant.

kev-cao commented on pkg/backup/schedule_exec.go:
```diff
@@ -362,6 +363,11 @@ func (e *scheduledBackupExecutor) backupSucceeded(
 		e.metrics.RpoMetric.Update(details.(jobspb.BackupDetails).EndTime.GoTime().Unix())
 	}
 
+	if details.(jobspb.BackupDetails).SpecificTenantIds != nil {
```
We have the `UpdatesLastBackupMetric` for determining the "special" schedule that will update the metric — should we do the same for these backup tenant schedules?

Thoughts @msbutler?

msbutler replied:
yeah, it looks like the update for this metric should be in the `UpdatesLastBackupmetric` branch above.

kev-cao commented on pkg/backup/schedule_exec.go:
```diff
@@ -362,6 +363,11 @@ func (e *scheduledBackupExecutor) backupSucceeded(
 	}
 
+	if details.(jobspb.BackupDetails).SpecificTenantIds != nil {
+		e.metrics.RpoTenantMetric.Update(map[string]string{"tenant_id": details.(jobspb.BackupDetails).SpecificTenantIds},
```
`SpecificTenantIDs` is a `[]string`, so you'll want to loop over them and emit a metric for each one.

kev-cao commented:
@edwardguo-crl I'm not aware of a way to use a custom Cockroach binary for the clusters in your local CC deployment, and my gut tells me that we don't have a mechanism built for that. @benbardin any insights on this?

msbutler commented:
@edwardguo-crl thanks for opening this! A few general process nits: 
- it looks like you're linking the crdb pr to a cc issue. In general, crdb pr's shoudl be linked to crdb issues. could you open a corresponding crdb issue instead?
- once you have the crdb issue, you can follow the general crdb commit guidelines: this pr should be named: `backup: export tenant ....`. The commit message should contain: Fixes {crdb_issue}.

msbutler commented:
> Is there a way I can test the exported metric before running it in staging?

@edwardguo-crl i would augment a test in backup_test.go that creates a backup of several tenants and verify the metrics were updated using some of the infra used in https://github.com/cockroachdb/cockroach/blob/master/pkg/sql/metric_test.go

edwardguo-crl committed:
backup: Export tenant backup schedule metric

Export Cloud tenant backup schedule completion metric with tenant id tag.

Issue: #141167

Epic: None

Release note: None

kev-cao commented on pkg/backup/schedule_exec.go:
```diff
@@ -358,8 +359,17 @@ func (e *scheduledBackupExecutor) backupSucceeded(
 
 	// If this schedule is designated as maintaining the "LastBackup" metric used
 	// for monitoring an RPO SLA, update that metric.
+	log.Infof(ctx, "args.UpdatesLastBackupMetric: %t", args.UpdatesLastBackupMetric)
 	if args.UpdatesLastBackupMetric {
 		e.metrics.RpoMetric.Update(details.(jobspb.BackupDetails).EndTime.GoTime().Unix())
+		log.Info(ctx, "backup succeeded")
```
Remove log statements

kev-cao commented on pkg/backup/schedule_exec.go:
```diff
@@ -578,6 +588,12 @@ func init() {
 					}),
+					RpoTenantMetric: metric.NewExportedGaugeVec(metric.Metadata{
+						Name:        "schedules.BACKUP.tenant-last-completed-time",
+						Help:        "The Unix timestamp of the most recently completed backup by host schedule for the tenant",
```
Something like "The unix timestamp of the most recently completed tenant backup by a host schedule specified as maintaining this metric" would be more fitting and similar to the existing description.

msbutler replied:
I like: "The unix timestamp of the most recently completed host scheduled backup by virtual cluster specified as maintaining this metric"

kev-cao commented on pkg/backup/schedule_exec_test.go:
```diff
@@ -0,0 +1,91 @@
+			RpoTenantMetric: metric.NewExportedGaugeVec(metric.Metadata{}, []string{"tenant_id"}),
+		},
+	}
+	env := scheduledjobs.ProdJobSchedulerEnv
```
It seems like our existing tests use a test environment instead of the production scheduler environment. I'm not familiar with this part of the codebase though, so @msbutler would have a better idea on how to best do the testing here.

msbutler replied:
I don't think the env matters here. @edwardguo-crl this is an awesome test!

msbutler commented on pkg/backup/schedule_exec.go:
```diff
@@ -574,10 +581,16 @@ func init() {
 					RpoMetric: metric.NewGauge(metric.Metadata{
 						Name:        "schedules.BACKUP.last-completed-time",
-						Help:        "The unix timestamp of the most recently completed backup by a schedule specified as maintaining this metric",
+						Help:        "The unix timestamp of the most recently completed tenant backup by a host schedule specified as maintaining this metric",
```
I would revert this change. The metric is bumped by a schedule within the cluster, so "host" is misleading. We also typically refer to backups taken _within_ a tenant as just a "backup". A "tenant backup" typically refers to a backup _of a_ tenant.

msbutler commented on pkg/backup/schedule_exec.go:
```diff
@@ -574,10 +581,16 @@ func init() {
 						Unit:        metric.Unit_TIMESTAMP_SEC,
 					}),
+					RpoTenantMetric: metric.NewExportedGaugeVec(metric.Metadata{
+						Name:        "schedules.BACKUP.tenant-last-completed-time",
```
nit: we're trying to scrub all mentions of "tenant" in our codebase and docs. Use virtual cluster instead.

msbutler replied:
how about this metric name: "schedules.BACKUP.last-completed-time-by-virtual_cluster"

msbutler commented on pkg/backup/schedule_exec_test.go:
```diff
@@ -0,0 +1,93 @@
+	return tenantIDs
+}
+
+func verifyRPOTenantMetricLabels(
```
could you also pass in and verify the expected end time?

msbutler commented on pkg/backup/schedule_exec_test.go:
```diff
@@ -0,0 +1,93 @@
+		err := executor.backupSucceeded(ctx, nil, schedule, details, env)
+		require.NoError(t, err)
+
+		expectedTenantIDs := []string{"system", "2"}
```
nit: also verify that `executor.metrics.RpoMetric.Value()` was also updated in this case?

edwardguo-crl replied:
I don't think there is a `Value()` method in RpoMetric. I verified the Gauge metric value in a separate function.

msbutler reviewed (commented):
@edwardguo-crl awesome job with the test! this patch is nearly there.

edwardguo-crl commented:
tftr!

bors r+
